source/common/network/tcp_listener_impl.cc

```diff
@@ -99,7 +99,7 @@ void TcpListenerImpl::setupServerSocket(Event::DispatcherImpl& dispatcher, Socke
 
   // Although onSocketEvent drains to completion, use level triggered mode to avoid potential
   // loss of the trigger due to transient accept errors.
-  file_event_ = socket.ioHandle().createFileEvent(
+  file_event_ = socket.ioHandle().createManagedFileEvent(
```

Comment:
Actually there is a gotcha here as I discovered from the integration tests. 

When we set `concurency_ = `2 in the integration tests it seems like there are two workers that both trying to create a tcp_listener on the same fd. Which with the new changes is crashing badly. From the conversation I had in the slack channel (see [here](https://envoyproxy.slack.com/archives/C78HA81DH/p1602695339118600?thread_ts=1602261758.081000&cid=C78HA81DH) this is the intended behavior for this listener

one example is at `IntegrationTest.PerWorkerStatsAndBalancing/IPv6`

Reply:
Added a comment in a38da8e68315fb2524f0654006b30ef4ca6bb42c

Reply:
By browsing through the code in `listener_impl` I see that UDS also share the socket

```
    ASSERT(local_address_->type() == Network::Address::Type::Pipe);
    // Listeners with Unix domain socket always use shared socket.
    create_socket = true;
```

Reply:
Another issue with having `SO_REUSEPORT` on by default is that it might break the hot_restart scenario on some older linux kernel releases.

See the comment at `listener.proto`

Reply:
I wonder if the non-SO_REUSE_PORT case can be fixed with the help of dup(2) to create multiple equivalent listening sockets.  The naming and documentation of the createManagedFileEvent method may help avoid us running into problems in the future.

Reply:
I can try adding a `deepCopy()` to `ioHandle()` that utilises `dup(2)` on Posix and `WSADuplicateSocket` on Windows to duplicate the handle correctly when we the socket

Reply:
Here https://github.com/envoyproxy/envoy/blob/master/source/server/listener_impl.cc#L129?

Reply:
I got the duplicate socket functionality working and now I think we don't need both of the functions. Just do the same thing for both tcp and udp listeners.

Reply:
We are good every test passes with duplicating the socket, resolving the comment